Add support for properties, JSON and YAML configuration in Core SDK

mraible opened the pull request:
This fixes #597 and #588 and #598. 

There's still more to do on 588 related to all the properties I added to `extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json` from `web.stormpath.properties`. Specifically, all the descriptions and default values still need to be populated.

I'm creating this PR now in hopes of getting it reviewed before I continue the tedious work of populating descriptions and default values. 😉

mrioan commented on extensions/servlet/src/test/groovy/com/stormpath/sdk/servlet/config/SpecConfigVersusWebPropertiesTest.groovy:
```diff
@@ -0,0 +1,99 @@
+ *
+ * @since 1.0.0
+ */
+class SpecConfigVersusWebPropertiesTest {
```
nice!

mrioan commented on extensions/servlet/src/test/groovy/com/stormpath/sdk/servlet/config/SpecConfigVersusWebPropertiesTest.groovy:
```diff
@@ -0,0 +1,99 @@
+            }
+        }
+
+        assertEquals diff.size(), 0, "Missing keys in default config: ${diff}"
```
Could you please add a clear message of why this error arose? Builds will start failing out of the blue as soon as someone changes the spec. Please add something like "It is very likely that a new property was added in XX and the Java SDK is missing it. It must be supported in the Java SDK in order for the build to pass.  In the meantime you could allow the build to pass by adding the missing property to the NOT_YET_IMPLEMENTED list but be sure to remember to addd support for it in the near term".

mrioan replied:
Could you please add a clear message of why this error arose? Builds will start failing out of the blue as soon as someone changes the spec. Please add something like "It is very likely that a new property was added in XX and the Java SDK is missing it. It must be supported in the Java SDK in order for the build to pass.  In the meantime you could allow the build to pass by adding the missing property to the NOT_YET_IMPLEMENTED list but be sure to remember to addd support for it in the near term".

mrioan replied:
Could you please add a clear message of why this error arose? Builds will start failing out of the blue as soon as someone changes the spec. Please add something like "It is very likely that a new property was added in XX and the Java SDK is missing it. It must be supported in the Java SDK in order for the build to pass.  In the meantime you could allow the build to pass by adding the missing property to the NOT_YET_IMPLEMENTED list but be sure to remember to addd support for it in the near term".

mrioan commented on extensions/servlet/src/test/groovy/com/stormpath/sdk/servlet/config/SpecConfigVersusWebPropertiesTest.groovy:
```diff
@@ -0,0 +1,99 @@
+            specProperties.containsKey(k) ? null : k
+        }
+
+        assertEquals diff.size(), 66, "Missing keys in spec config: ${diff}"
```
Could you please add a clear message of why this error arose? Builds will start failing out of the blue as soon as someone changes the spec. Please add something like "It is very likely that a new property was added in XX and the Java SDK is missing it. It must be supported in the Java SDK in order for the build to pass.  In the meantime you could allow the build to pass by adding the missing property to the NOT_YET_IMPLEMENTED list but be sure to remember to addd support for it in the near term".

mrioan commented on extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy:
```diff
@@ -0,0 +1,58 @@
+import org.testng.annotations.Test
+
+import static org.testng.Assert.assertEquals
+/**
```
Blank line please

mraible replied:
Just so I know what you're asking - you're saying there should be a blank line before the javadoc, correct?

mrioan commented on extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy:
```diff
@@ -0,0 +1,58 @@
+ * @since 1.0.0
+ */
+class SpringMetadataVersusCorePropertiesTest {
+    def METADATA_CONFIG_LOCATION = ClasspathResource.SCHEME_PREFIX + "META-INF/additional-spring-configuration-metadata.json"
```
idem here

mraible replied:
Blank line after the class?

mrioan commented on extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy:
```diff
@@ -0,0 +1,58 @@
+/**
+ * A test that compares the properties in com/stormpath/sdk/config/stormpath.properties
+ * with the ones defined in additional-spring-configuration-metadata.json.
+ *
```
Could you please addd a comment saying something like "This test will alert when a new property has been added to the framework but it was not documented in `additional-spring-configuration-metadata.json`" ?

mrioan commented on extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json:
```diff
@@ -1,6 +1,9 @@
     {
+      "name": "stormpath.application"
+    },
+    {
```
I think these properties should be documented, should't them?

mraible replied:
I'm not sure if the groups have a `description` field. @lhazlewood I assumed these groups were only necessary for keys that had a number of properties under the same "group". Is that correct?

mrioan commented on extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json:
```diff
@@ -482,22 +473,899 @@
+    },
+    {
+      "name": "stormpath.application.href",
+      "type": "java.lang.String",
```
Missing description and default value in some of these...

mraible replied:
You are correct. I copied the descriptions and defaultValues from `web.stormpath.properties` (if they existed). @dogeared added issue #620 for the rest.

mrioan replied:
Great

mrioan commented on impl/src/main/java/com/stormpath/sdk/impl/config/ClientConfiguration.java:
```diff
@@ -0,0 +1,160 @@
+ * @since 1.0
+ */
+public class ClientConfiguration {
+    private String apiKeyFile;
```
blank line please

mrioan commented on impl/src/main/java/com/stormpath/sdk/impl/config/ClientConfiguration.java:
```diff
@@ -0,0 +1,160 @@
+
+/**
+ * This class holds the default configuration properties
+ *
```
Is this accurate? I would think that this class is a placeholder for all the configuration properties, both default ones and overwritten by the user. If so I would add a comment saying something like "During application initialization all the properties found in the pre-defined locations that are defined by the user will be added here in the order defined in XXX. Unset value will keep default values pre-defined in YYY"

XXX: is a link to either the spec, the github issue  or the wiki page that clearly is documenting this order
YYY: link to the file in the Java SDK containing the default values

mrioan commented:
I would like to see a few tests that demonstrate that the overrides are happening in the right order.

Back to you @mraible

mraible commented:
@mrioan I fixed Javadoc issues, improved test failure messages, fixed the syntax issues you mentioned and improved [one test](https://github.com/stormpath/stormpath-sdk-java/pull/617/commits/c8fd0d7e0ab15324bb031b988e5683a5ef67c271) to prove overrides are working. If you'd like additional tests, please let me know. 

I could create `~/stormpath.*` files on-the-fly for more rigorous tests. Unfortunately, I've found it difficult to delete these files while the tests are running - so other tests will pick up this configuration. That's why I didn't do it this way.

mrioan commented:
@mraible LGTM. If you think @lhazlewood or @dogeared should take a look at it as well for some reason then contact them. Otherwise you are good to merge. Back to you. Thanks
